### PR TITLE
Drill 2517: Apply directory-based partition pruning before reading files in planning.

### PR DESCRIPTION
**Pull request by jinfengni**
1. Run the pre-commit tests and unit tests. Some queries in pre-commit suites have changed plan. Most of the changed plan looks better than before. The only exception is for the cases caused by one existing issue (DRILL-4279), when * column is used together with SKIP_ALL mode. That happens when the filter is applied and then removed, for the following query:

SELECT count(*) from T1 where dir0 = 1990 and dir1 = 'Q1'.

2. I need figure out how to cherry-pick Adam's patch in DRILL-2517, since that's the initial work on this issue, although there is quite big change from that patch.  

@amansinha100 , could you please take a look and give some initial review comments? Thanks!

### CODE REVIEW

File: exec/java-exec/src/main/java/org/apache/drill/exec/planner/logical/partition/PruneScanRule.java [hunk 1]
```diff
@@ -78,71 +94,68 @@ public PruneScanRule(RelOptRuleOperand operand, String id, OptimizerRulesContext
     this.optimizerContext = optimizerContext;
   }
 
-  public static final RelOptRule getFilterOnProject(OptimizerRulesContext optimizerRulesContext) {
-    return new PruneScanRule(
-        RelOptHelper.some(DrillFilterRel.class, RelOptHelper.some(DrillProjectRel.class, RelOptHelper.any(DrillScanRel.class))),
-        "PruneScanRule:Filter_On_Project",
-        optimizerRulesContext) {
+  private static class DirPruneScanFilterOnProjectRule extends PruneScanRule {
+    private final boolean calciteLogical;
```
Comment (amansinha100) [inline exec/java-exec/src/main/java/org/apache/drill/exec/planner/logical/partition/PruneScanRule.java]: It's not clear to me why a flag is needed to keep track of Calcite logical...ideally this should be handled by the type of the RelNode. We may need to do some refactoring of this rule to separate out the ones that apply during Calcite logical planning vs. Drill logical planning.

File: exec/java-exec/src/main/java/org/apache/drill/exec/planner/FileSystemPartitionDescriptor.java [hunk 2]
```diff
@@ -84,9 +99,17 @@ public int getMaxHierarchyLevel() {
...
   public GroupScan createNewGroupScan(List<String> newFiles) throws IOException {
-    final FileSelection newSelection = FileSelection.create(null, newFiles, getBaseTableLocation());
-    final FileGroupScan newScan = ((FileGroupScan)scanRel.getGroupScan()).clone(newSelection);
-    return newScan;
+    if (scanRel instanceof DrillScanRel) {
+      final FileSelection newFileSelection = new FileSelection(null, newFiles, getBaseTableLocation());
+      final FileGroupScan newScan = ((FileGroupScan)((DrillScanRel)scanRel).getGroupScan()).clone(newFileSelection);
+      return newScan;
+    } else {
+      throw new UnsupportedOperationException("Does not allow to get groupScan for EnumerableTableScan");
```
Comment (amansinha100) [inline exec/java-exec/src/main/java/org/apache/drill/exec/planner/FileSystemPartitionDescriptor.java]: The else condition could be anything other than DrillScanRel, so you need not mention EnumerableTableScan.

File: exec/java-exec/src/main/java/org/apache/drill/exec/planner/logical/partition/PruneScanRule.java (hunk 1 above)
Comment (jinfengni) [inline exec/java-exec/src/main/java/org/apache/drill/exec/planner/logical/partition/PruneScanRule.java]: You are right. The flag is not needed now. I will refactor the rules.

File: exec/java-exec/src/main/java/org/apache/drill/exec/planner/FileSystemPartitionDescriptor.java [hunk 3]
```diff
@@ -82,11 +101,19 @@ public int getMaxHierarchyLevel() {
     return MAX_NESTED_SUBDIRS;
   }
 
-  @Override
-  public GroupScan createNewGroupScan(List<String> newFiles) throws IOException {
-    final FileSelection newSelection = FileSelection.create(null, newFiles, getBaseTableLocation());
-    final FileGroupScan newScan = ((FileGroupScan)scanRel.getGroupScan()).clone(newSelection);
-    return newScan;
+//  @Override
+//  public GroupScan createNewGroupScan(List<String> newFiles) throws IOException {
```
Comment (amansinha100) [inline exec/java-exec/src/main/java/org/apache/drill/exec/planner/FileSystemPartitionDescriptor.java]: This function can be removed completely...

File: exec/java-exec/src/main/java/org/apache/drill/exec/planner/logical/partition/PruneScanRule.java [hunk 4]
```diff
@@ -382,5 +381,28 @@ protected OptimizerRulesContext getOptimizerRulesContext() {
...
+  public abstract PartitionDescriptor getPartitionDescriptor(PlannerSettings settings, TableScan scanRel);
+
+  private static boolean isQualifiedDirPruning(final TableScan scan) {
+    if (scan instanceof EnumerableTableScan) {
+      DrillTable drillTable;
+      drillTable = scan.getTable().unwrap(DrillTable.class);
+      if (drillTable == null) {
+        drillTable = scan.getTable().unwrap(DrillTranslatableTable.class).getDrillTable();
+      }
+      final Object selection = drillTable.getSelection();
```
Comment (amansinha100) [inline exec/java-exec/src/main/java/org/apache/drill/exec/planner/logical/partition/PruneScanRule.java]: Would be good to assert drillTable != null or put an if() condition.

**Comment by amansinha100** [comment]
Revised patch looks good to me..other than a few minor comments.   +1.

**Comment by jinfengni** [comment]
this PR has been merged to drill 1.6.0. Close this PR.

**Pull request closed without merging**